#3374 If path name is very long then pytest will split name for several lines

feuillemorte opened the pull request:
Fixes #3374

Long names:

![selection_023](https://user-images.githubusercontent.com/9434344/38578894-f3788908-3d0d-11e8-8597-b1ae91a6f865.png)

Long name full line:
![selection_025](https://user-images.githubusercontent.com/9434344/38578915-01ccc0c8-3d0e-11e8-8091-cca760d07609.png)

Long name full line + 1 letter:
![selection_026](https://user-images.githubusercontent.com/9434344/38578937-0c514e74-3d0e-11e8-9175-93403cd191da.png)

Very beautiful when progress prints in every line but it is a problem with progress count: it counts only after run a test (after report write a dot), so, there is a problem:
![selection_024](https://user-images.githubusercontent.com/9434344/38578991-47ea50c0-3d0e-11e8-8bda-876c4ba741ee.png)

nicoddemus commented:
Thanks @feuillemorte for the PR!

Personally I prefer to just write the full path name and let the terminal wrap that around for us if needed, otherwise this makes it hard to copy/paste the test name which is useful on occasion:

```
tests/test_very_long_name_test_very_long_name_test_very_long_name_test_very_long_nam
e_test_very_long_name.py ..                                                   [ 50%]
```

If we write to the edge of the screen in a way that if we would mess with the alignment of the progress indicator, we should start a new line instead. Using the example from #3374:

```
tests\test_general\test_find_last_operator\test_correct_expressions.py . [ 77%]
.......                                                                  [ 93%]
tests\test_general\test_find_last_operator\test_incorrect_expressions.py 
.                                                                        [ 95%]
tests\test_general\test_find_last_operator\test_monomials.py ..          [100%]
```

What do others think? 

cc @InDevRus

feuillemorte commented:
@nicoddemus 
I fixed your comment and now it looks more convenient. Also I deleted test because I don't know how to set terminal width. If you can give me advice how to write a test, please, do it.

Look at the pictures how it works now:

Long names:
![selection_035](https://user-images.githubusercontent.com/9434344/38742572-a3746f82-3f45-11e8-884a-64bd481e69a8.png)

Long name full line:
![selection_036](https://user-images.githubusercontent.com/9434344/38742597-b04240ae-3f45-11e8-82cb-1203288977b9.png)

Long name full line + 1 symbol:
![selection_037](https://user-images.githubusercontent.com/9434344/38742610-bb2f3882-3f45-11e8-914d-d5628313e490.png)

But I found a little problem here:
![selection_038](https://user-images.githubusercontent.com/9434344/38742622-c4a88648-3f45-11e8-97b2-2ee107514805.png)

I have 99% at 2 lines and 0% at the beginning. Is it a bug?

feuillemorte commented:
Fixed some bug:

before:
![selection_039](https://user-images.githubusercontent.com/9434344/38744285-12e2c810-3f4a-11e8-9efc-41226999c965.png)

after:
![image](https://user-images.githubusercontent.com/9434344/38744315-25e1045e-3f4a-11e8-8d29-49eeb136894b.png)
![selection_040](https://user-images.githubusercontent.com/9434344/38744325-2b9e9ee2-3f4a-11e8-82cf-042658ab088c.png)

nicoddemus commented on _pytest/terminal.py:
```diff
@@ -216,6 +217,11 @@ def write_fspath_result(self, nodeid, res):
             self._tw.line()
             self._tw.write(fspath + " ")
+            width = self._tw.fullwidth - self._tw.chars_on_current_line
+            if 0 < width < 10 or self._tw.chars_on_current_line % self._screen_width > self._screen_width - 10:
```
Why `10`? I suggest using a variable with a descriptive name instead. The same can be done for the entire conditional, for example:

```python
something_descriptive =  0 < width < 10 or self._tw.chars_on_current_line % self._screen_width > self._screen_width - 10
```

Also can you explain this logic here?

feuillemorte replied:
10 symbols because of this line:
```.py [ 80%]```
this is magic 10 symbols :)

logic:
if you have long name and remaining wight less then 10 symbols, then add new line.
OR
if you have long name (3 lines) and width of tail of this name more then screen width - 10 symbols, then also add new line

nicoddemus replied:
> logic:
if you have long name and remaining widht less then 10 symbols, then add new line.
OR
if you have long name (3 lines) and width of tail of this name more then screen width - 10 symbols, then also add new line 

Could you please separate each condition in their own variable with descriptive name? A descriptive variable name often is enough to explain a complex situation and avoid having to write comments or explain it to people.

nicoddemus replied:
> How I can write test on it? Does it have a value like
tempdir.runtests(--screen-width=120)
?

I don't think so, but I believe `testdir` uses a hard coded value internally. You can try to match against the entire line instead of using glob wildcards.

feuillemorte replied:
```
I believe testdir uses a hard coded value internally.
```
I tried to run my test and it failed because this value was changed and my test with long name on 3 lines become to test with 1 line. So that's why I deleted it.

Also, could you please some test on your terminal after my fix and before merging? I'm some afraid of my magic constants :) It looks nice in my terminals (I tried 2 terminals), but who knows how it will not on my machine.

nicoddemus replied:
Hey @feuillemorte, did you push more commits? I don't see them here.

feuillemorte replied:
@nicoddemus I fixed your comments, but I can't push because I can't to write a test. I try to get an object of terminal and get screen width. It says me 80 symbols, but actually it's wrong value inside runpytest() function (159 symbols):
(sorry, I can't paste image because some amazon ip's blocked in Russia, please, look url)
```
test_01_very_long_name_very_long_name_very_long_name_very_long_name_very_long_name_very_long_name_very_l="""
                import pytest
                from _pytest.terminal import TerminalReporter
                term = TerminalReporter(pytest.config)
                assert term._screen_width == 1
                
                @pytest.mark.parametrize('i', range(10))
                def test_bar(i): assert 1
            """,
```
```
test_01_very_long_name_very_long_name_very_long_name_very_long_name_very_long_name_very_long_name_very_l.py:4: in <module>
    assert term._screen_width == 1
E   assert 80 == 1
```
output:
https://pp.userapi.com/c824503/v824503903/11fc2b/8XBN-QcJDh0.jpg

feuillemorte replied:
@nicoddemus I pushed some code but without tests

nicoddemus commented on _pytest/terminal.py:
```diff
@@ -355,7 +361,10 @@ def pytest_runtest_logfinish(self, nodeid):
             else:
                 past_edge = self._tw.chars_on_current_line + self._PROGRESS_LENGTH + 1 >= self._screen_width
-                if past_edge:
+                width = self._tw.chars_on_current_line % self._screen_width + 9
```
Again a magic constant, please use a named constant instead. 😁

nicoddemus replied:
I see `9` should be `_MAX_LEN_LEFT - 1` if we add a constant for it, is that right?

nicoddemus commented:
Hi @feuillemorte, 

It has been a long time since it has last seen activity, plus we have made sweeping changes on `master` to drop Python 2.7 and 3.4 support, so this PR has some conflicts which require attention.

In order to clear up our queue and let us focus on the active PRs, I'm closing this PR for now.

Please don't consider this a rejection of your PR, we just want to get this out of sight until you have the time to tackle this again. If you get around to work on this in the future, please *don't hesitate* in re-opening this!

Thanks for your work, the team definitely appreciates it!
